File: backend/app/schemas/export.py
# backend/app/schemas/export.py
from typing import Literal
from pydantic import BaseModel, model_validator


class ExportRequest(BaseModel):
    scope: Literal["image", "batch", "all"]
    imageId: int | None = None
    batchId: int | None = None
    formats: list[Literal["mask", "coco", "labelme"]]
    skipUnconfirmed: bool = False

    @model_validator(mode="after")
    def _check_ids(self):
        if self.scope == "image" and self.imageId is None:
            raise ValueError("imageId is required when scope is 'image'")
        if self.scope == "batch" and self.batchId is None:
            raise ValueError("batchId is required when scope is 'batch'")
        if not self.formats:
            raise ValueError("formats must not be empty")
        return self


class PendingItem(BaseModel):
    image: str
    labels: list[str]


class ExportResponse(BaseModel):
    exportDir: str
    imageCount: int
    annotationCount: int
    maskCount: int
    pending: list[PendingItem] = []
    errors: list[dict[str, str]] = []

File: frontend/src/api/export.ts
import { apiClient, ApiError } from './client';

export type ExportScope = 'image' | 'batch' | 'all';
export type ExportFormat = 'mask' | 'coco' | 'labelme';

export interface ExportRequest {
  scope: ExportScope;
  imageId?: number | null;
  batchId?: number | null;
  formats: ExportFormat[];
  skipUnconfirmed?: boolean;
}

export interface PendingItem {
  image: string;
  labels: string[];
}

export interface ExportError {
  file: string;
  error: string;
}

export interface ExportResponse {
  exportDir: string;
  imageCount: number;
  annotationCount: number;
  maskCount: number;
  pending: PendingItem[];
  errors: ExportError[];
}

export function unconfirmedPending(err: unknown): PendingItem[] | null {
  if (!(err instanceof ApiError) || err.status !== 409) return null;
  const detail = (err.body as { detail?: { code?: string; pending?: PendingItem[] } } | null)?.detail;
  if (detail && detail.code === 'unconfirmed_labels' && Array.isArray(detail.pending)) {
    return detail.pending;
  }
  return null;
}

export async function runExport(body: ExportRequest): Promise<ExportResponse> {
  return apiClient.post<ExportResponse>('/export', body);
}
